numpy/core/src/umath/loops.c.src

```diff
@@ -840,28 +843,83 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 void
     UNARY_LOOP_FAST(@type@, @type@, *out = in > 0 ? 1 : (in < 0 ? -1 : 0));
 }
 
+/* Libdivide only supports 32 and 64 bit types
+ * We try to pick the best possible one */
+#if NPY_BITSOF_@TYPE@ <= 32
+#define libdivide_@type@_t libdivide_s32_t
+#define libdivide_@type@_gen libdivide_s32_gen
+#define libdivide_@type@_do libdivide_s32_do
+#else
+#define libdivide_@type@_t libdivide_s64_t
+#define libdivide_@type@_gen libdivide_s64_gen
+#define libdivide_@type@_do libdivide_s64_do
+#endif
+
 NPY_NO_EXPORT void
 @TYPE@_divide(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
 {
-    BINARY_LOOP {
-        const @type@ in1 = *(@type@ *)ip1;
+    BINARY_DEFS
+
+    /* When the divisor is a constant, use libdivde for faster division */
+    if (steps[1] == 0) {
         const @type@ in2 = *(@type@ *)ip2;
-        /*
-         * FIXME: On x86 at least, dividing the smallest representable integer
-         * by -1 causes a SIFGPE (division overflow). We treat this case here
-         * (to avoid a SIGFPE crash at python level), but a good solution would
-         * be to treat integer division problems separately from FPU exceptions
-         * (i.e. a different approach than npy_set_floatstatus_divbyzero()).
-         */
-        if (in2 == 0 || (in1 == NPY_MIN_@TYPE@ && in2 == -1)) {
-            npy_set_floatstatus_divbyzero();
-            *((@type@ *)op1) = 0;
-        }
-        else if (((in1 > 0) != (in2 > 0)) && (in1 % in2 != 0)) {
-            *((@type@ *)op1) = in1/in2 - 1;
+
+        /* If divisor is 0, we need not compute anything*/
+        if (in2 == 0) {
+            BINARY_LOOP_SLIDING {
+                npy_set_floatstatus_divbyzero();
```

Comment:
Thanks a lot for catching this :) . I have done a 0/Nat check at the start and skipping looping with `n == 0` check. Although I am curious, will the `n==0 `check really matter? We are not allocating any memory yeah?

Reply:
You are always reading one element from the second argument. Almost always, that will probably just work, but f there are 0 items to be iterated, you are not allowed to read one.

Reply:
Ohh ok, I get it now, thanks, so yeah what I have currently must change, the n==0 must be above all others.

Reply:
>  i.e. please add a test for np.array([], dtype=np.intp) // 0 giving no warning
Also, this should give a warning right? Or its not needed?

Reply:
Yes, and no it should not give a warning, since no division is ever performed.

Reply:
Thanks, I have one more doubt: https://github.com/numpy/numpy/blob/94b347e867759325d8586ad3d1c3e01c590d83c1/numpy/core/src/umath/loops.c.src#L1358-L1359
Should we not set `npy_set_floatstatus_divbyzero` here, at least for `in2==0`?

Reply:
If `in1` is NPY_DATETIME_NAT` no divbyzero should be set.  If it isn't it should probably bet set, but if current master doesn't do it (I am not sure), then we might as well do that in a followup.

Reply:
The rule for NaNs is that that you should always set the flag/give a warning when a new NaN is created and never when the input was NaN to begin with (normally, for floats there are also "signalling NaNs" which behave differently, but we don't really handle those).  Not sure if you could argue that if you replace the NaT with any other value it should also warn, but I don't think we would set it in that case currently (but if IEEE disagrees I will change my mind on a dime).

Reply:
> The rule for NaNs is that ...

That makes sense. We should [add it to docs](https://numpy.org/devdocs/user/misc.html#how-numpy-handles-numerical-exceptions)

Reply:
Oh I see. So coupling:
> set the flag/give a warning when a new NaN

and

https://github.com/numpy/numpy/blob/94b347e867759325d8586ad3d1c3e01c590d83c1/numpy/core/src/umath/loops.c.src#L1441-L1443

We can set the value when `in2` is 0, and not do anything when `in1` is Nat?

Reply:
Again I think, we need to have a look at this anyways:
https://github.com/numpy/numpy/blob/94b347e867759325d8586ad3d1c3e01c590d83c1/numpy/core/src/umath/loops.c.src#L1437-L1439

We need not set `npy_set_floatstatus_invalid` here yeah, as input is NaT?

Reply:
Jupp, you are totally right on both accounts as far as I see (I think I did the same here: https://github.com/numpy/numpy/pull/16458) but never finished that up :(.

I would be happy to just retain it for this iteration and follow up, but if you prefer you can fix it (with new tests).

Reply:
Another point to note in the above code is when we generate NaN, as per [wiki](https://en.wikipedia.org/wiki/NaN), 
> Most operations with at least one NaN operand.

So setting to 0 seems like something we need to probe.

Reply:
I would like to take this as a follow-up PR as we need to make two major changes:
For both timedeltas, we need to do this: https://github.com/numpy/numpy/blob/000dfda49a012bd38efc6fbb89ece58279895bad/numpy/core/src/umath/loops.c.src#L1446-L1448

But as I mentioned in https://github.com/numpy/numpy/pull/17727#issuecomment-726084647, we are truncating instead of flooring. Second being the setting of invalid division.

Reply:
The truncating may be on purpose (although correct rounding would be right!). I think that is the normal division operator, not the floor division operator.

Anyway, if you want to be thorough, can you open a new issue about this. This is already a lot of disucssion here!

Reply:
Sure @seberg, I have created #17831.